docker/docker_client.go

```diff
@@ -244,37 +248,96 @@ func newDockerClient(sys *types.SystemContext, registry, reference string) (*doc
 	if err := tlsclientconfig.SetupCertificates(certDir, tr.TLSClientConfig); err != nil {
 		return nil, err
 	}
+	tr.TLSClientConfig.InsecureSkipVerify = sysContextSkipVerify(sys)
+
+	// Find the registry
+	reg, err := sysregistriesv2.FindRegistry(sys, reference)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error loading registries")
+	}
+	if reg != nil {
+		// Retrieve the docker client from the found registry
+		return dockerClientFromSysRegistry(ctx, sys, reg, tr), nil
+	}
 
+	// No registry found within the configuration, fallback to the provided
+	// one
+	return &dockerClient{
+		registry:              registry,
+		sys:                   sys,
+		client:                &http.Client{Transport: tr},
+		insecureSkipTLSVerify: tr.TLSClientConfig.InsecureSkipVerify,
+	}, nil
+}
+
+func sysContextSkipVerify(sys *types.SystemContext) bool {
 	// Check if TLS verification shall be skipped (default=false) which can
 	// either be specified in the sysregistriesv2 configuration or via the
 	// SystemContext, whereas the SystemContext is prioritized.
-	skipVerify := false
 	if sys != nil && sys.DockerInsecureSkipTLSVerify != types.OptionalBoolUndefined {
 		// Only use the SystemContext if the actual value is defined.
-		skipVerify = sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
-	} else {
-		reg, err := sysregistriesv2.FindRegistry(sys, reference)
-		if err != nil {
-			return nil, errors.Wrapf(err, "error loading registries")
+		return sys.DockerInsecureSkipTLSVerify == types.OptionalBoolTrue
+	}
+	return false
+}
+
+// dockerClientFromSysRegistry returns a `*dockerClient` for the provided
+// sysregistriesv2. If the registry contains any mirrors, it will be tried
+// to detect their properties. The mirror will be used if the property
+// detection succeeds. If not, the fallback registry URL will be used instead.
+func dockerClientFromSysRegistry(
+	ctx context.Context,
+	sys *types.SystemContext,
+	registry *sysregistriesv2.Registry,
+	transport *http.Transport,
+) *dockerClient {
+	// registry != nil should already be checked by referring functions
+	// Check for available mirrors
+	for _, mirror := range registry.Mirrors {
+		// Adapt the InsecureSkipVerify only when not already set from the
+		// system context
+		if !transport.TLSClientConfig.InsecureSkipVerify {
+			transport.TLSClientConfig.InsecureSkipVerify = mirror.Insecure
+		}
+
+		client := &dockerClient{
+			client:                &http.Client{Transport: transport},
+			insecureSkipTLSVerify: mirror.Insecure,
+			registry:              mirror.URL,
+			sys:                   sys,
 		}
-		if reg != nil {
-			skipVerify = reg.Insecure
+
+		// Test the registries availability
+		if client.detectProperties(ctx) != nil {
+			logrus.Warnf("mirror '%v' does seems to be unavailable", mirror.URL)
```

Comment:
(Eventually, please use correct capitalization.

Also, I don’t think library code in c/image should just write to standard output like this — either use `logrus.Debug`*, or, if this information is really necessary (it might be), we will have to implement new kind of feedback channel (that somehow integrates with the `copy.Image` reporting and progress bars).)

Reply:
Yes, I agree that this is useful information that the user should be able to obtain, but that needs to happen through Go data structures.